### PR TITLE
global/consistent policy variable interpolation

### PR DESCRIPTION
**Pull request by kapilt**
addresses #2330 

todo, also see if we can address the cli passing of variables per #938

### CODE REVIEW

**Comment by kapilt** [comment]
not ready for merge till it addresses #2400

**Comment by kapilt** [comment]
@beastie29a thanks for trying out the branch, I've made a few adjustments to where variable expansion was taking place. It works ootb for your example now.

File: c7n/policy.py [hunk 1]
```diff
@@ -728,12 +715,56 @@ def is_lambda(self):
...
+            'bucket_region': '{bucket_region}',
+            'bucket_name': '{bucket_name}',
+            'source_bucket_name': '{source_bucket_name}',
+            'target_bucket_name': '{source_bucket_name}',
+            'target_prefix': '{target_prefix}',
+            'LoadBalancerName': '{LoadBalancerName}'
+        }
+
+    def expand_variables(self, variables):
+        """expand variables in the mode role and output_dir fields.
```
Comment (kapilt) [inline c7n/policy.py]: updated doc string

File: c7n/policy.py [hunk 2]
```diff
@@ -728,12 +715,56 @@ def is_lambda(self):
...
+        return {
+            # standard runtime variables for interpolation
+            'account': '{account}',
+            'account_id': self.options.account_id,
+            'region': self.options.region,
+            # non-standard runtime variables from local filter/action vocabularies
+            #
+            # notify action
+            'policy': self.data,
+            'event': '{event}',
```
Comment (kapilt) [inline c7n/policy.py]: this unfortunately may have some compatibility considerations, accessing keys of dictionaries in the format string won't work across this deferral mechanism where we pass in a simple string.

Comment (kapilt) [reply c7n/policy.py]: the use cases for this seem pretty minimal, given its for notify and in notify you can use a full template engine with access to the same variables.

**Comment by kapilt** [comment]
there's some odd interactions with format() on templates with embedded jinja. For the notify action, we currently support doing rendering of the subject line with jinja.

```python
>>> subject = '{{account}} has an issue in {{region}} with {{policy.name}}'
>>> subject.format()
'{account} has an issue in {region} with {policy.name}'
```

its unclear what options we have outside of handling this as a special case for preserving compatibility.

**Comment by kapilt** [comment]
addressed the notify subject issue via ugly hack, not sure there's a better option, tbh.

File: c7n/policy.py [hunk 3]
```diff
@@ -728,12 +715,68 @@ def is_lambda(self):
...
+            'now': utils.FormatDate(datetime.utcnow()),
+            # account increase limit action
+            'service': '{service}',
+            # s3 set logging action :-( see if we can revisit this one.
+            'bucket_region': '{bucket_region}',
+            'bucket_name': '{bucket_name}',
+            'source_bucket_name': '{source_bucket_name}',
+            'target_bucket_name': '{source_bucket_name}',
+            'target_prefix': '{target_prefix}',
+            'LoadBalancerName': '{LoadBalancerName}'
```
[the remark this answers is not in the document]
Comment (kapilt) [reply c7n/policy.py]: load balancer name was a key used in app elb `set-s3-logging` action as a var for the s3 prefix construction. basically its here as i'm trying to carry forward all those existing usages for compatibility.

File: tests/test_policy.py [hunk 4]
```diff
@@ -287,6 +287,26 @@ def test_policy_region_expand_global(self):
...
+        p = self.load_policy({
+            'name': 'compute',
+            'resource': 'aws.ec2',
+            'mode': {
+                'type': 'config-rule',
+                'role': 'arn:iam::{account_id}/role/FooBar'},
+            'actions': [
+                {'type': 'tag',
+                 'value': 'bad monkey {account_id} {region} {now:+2d%Y-%m-%d}'}
+            ]}, config={'account_id': '12312311', 'region': 'zanzibar'})
```
[the remark this answers is not in the document]
Comment (kapilt) [reply tests/test_policy.py]: internally load_policy will just merge it with a c7n.config object, there's not really a differential there afaics

File: c7n/policy.py [hunk 5]
```diff
@@ -764,13 +750,69 @@ def is_lambda(self):
...
+            'action_date': '{action_date}',
+            # tag action pyformat-date handling
+            'now': utils.FormatDate(datetime.utcnow()),
+            # account increase limit action
+            'service': '{service}',
+            # s3 set logging action :-( see if we can revisit this one.
+            'bucket_region': '{bucket_region}',
+            'bucket_name': '{bucket_name}',
+            'source_bucket_name': '{source_bucket_name}',
+            'target_bucket_name': '{source_bucket_name}',
```
Comment (tjstansell) [inline c7n/policy.py]: shouldn't this be `{target_bucket_name}`?

Comment (kapilt) [reply c7n/policy.py]: yup, good catch, thanks

**Comment by eamonnfaherty** [comment]
Does this branch mean we can use account_id in the period events as requested in this issue: https://github.com/cloud-custodian/cloud-custodian/issues/3155

**Comment by kapilt** [comment]
this pr means you can use {account_id} anywhere in a policy file, the issue in #3155 is unrelated to being able to specify {account_id}, if you have a question about that issue its best to ask there.

**Pull request merged**